### PR TITLE
Nightly CI runs

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------
As always, apologies for the size of the PR :) (the diff is mostly due to linting tho)

- Moved a bunch of `standalone_tests` to `yea` to be executed in the cloud/on specialized hardware.
  - This uses the relay functionality of `yea`. The tests are logged to the cloud app under the `wandb-sdk` entity/team with a service account (so this can be used for our pre-release activities).
  - We spin up a `k8s` cluster on GKE per nightly/standalone run and execute `standalone_tests` on it.
  - Tests are sharded as follows: `cpu`, `gpu`, `tpu`, and `local`. We build a dedicated Docker image per shard that is then spun up in a pod that is deployed to the cluster. 
  - The `WANDB_API_KEY` is propagated into pod specs with some shell magic, hope ppl appreciate it :)
  - Care was taken about the available hardware resources on the cluster vs what the pods request (and what the imposed limits are -- check out the pod specs).
  - Turned on linting for the standalone tests (`black` + `flake8`), hence so many changed filed.
  - The DAG on Circle looks like this:
image

- Added a `nightly` workflow to Circle CI.
  - Uses Circle's [Scheduled Pipelines](https://circleci.com/docs/2.0/scheduled-pipelines/). The trigger is configured [here](https://app.circleci.com/settings/project/github/wandb/client/triggers).
- Added (configurable) Slack integration. Standalone/nightly jobs post to `#sdk-ci-nightly`.
- Cleaned up `tox.ini` a bit.

Notes:
- A scheduled pipeline can sometimes stop running 😭. See [here](https://support.circleci.com/hc/en-us/articles/360059279412-Scheduled-workflows-did-not-run) re: things to try to restore it. Because of this possibility, I added a Slack notification on a nightly run start.
- Think about which features of the current regression suite to transfer to the nightly tests in the cloud.

TODOs: 
- Write a Notion doc with the detailed description of this saga.
- The `tpu` and `local` shards need to actually be implemented :) -- but that could be done in a future PR
- Move the remaining standalone tests to `yea` -- but that could also be done in a future PR. The GPU container/shard has been verified to make both `tf` and `pytorch` happy (the see the GPUs).
- Solve codecov! Currently, we do not account for coverage coming from this path.
- ~~Fix/improve the container status checks with the shell~~:
image

Testing
-------
So meta! On the CI ;)

### CODE REVIEW

File: standalone_tests/shards/gke_gpu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    && chown -R sdk:sdk /wandb
+RUN echo "sdk ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/90-sdk
+USER sdk
+ENV HOME=/home/sdk
+RUN chmod 777 /home/sdk
+
+# clone wandb/client repository and install wandb sdk
+RUN git clone https://github.com/wandb/client.git /wandb/client \
+    && cd /wandb/client \
+    && git checkout ${GIT_BRANCH} \
```
Comment (kptkin) [inline standalone_tests/shards/gke_gpu/Dockerfile:47]: just curious why do you need to clone? why not just install from repository?

File: standalone_tests/basic.yea [hunk 2]
```diff
@@ -0,0 +1,16 @@
+id: 0.standalone.basic
+tag:
+  shards:
+    - standalone-cpu
+    - standalone-gpu
```
Comment (kptkin) [inline standalone_tests/basic.yea:5]: ah nice we can run in multiple shards

File: .circleci/config.yml [hunk 3]
```diff
@@ -92,385 +108,894 @@ commands:
...
+#            TPU support
+#           gcloud beta container clusters update cluster-name --enable-tpu
+#           gcloud container clusters get-credentials
+#            TPU support
+#            --release-channel=stable \
+#            --scopes=cloud-platform \
+#            --enable-ip-alias \
+#            --enable-tpu
+#            e2-medium
+#            --accelerator=type=nvidia-tesla-k80,count=1
```
Comment (kptkin) [inline .circleci/config.yml:207]: commented out code? for future changes?

File: .circleci/config.yml [hunk 4]
```diff
@@ -92,385 +108,894 @@ commands:
...
+            pip install tox==<< pipeline.parameters.tox_version >>
+      - restore-tox-cache
+      - run:
+          name: Run tests
+          command: |
+            tox -v -e << parameters.toxenv >>
+          no_output_timeout: 25m
+      - save-tox-cache
+      - save-test-results
+      # conditionally post a notification to slack if the job failed
```
Comment (kptkin) [inline .circleci/config.yml:264]: why not publish also on success? If someone wants to look at the completed run?

File: .circleci/config.yml [hunk 5]
```diff
@@ -92,385 +108,894 @@ commands:
...
+        type: string
+      image_name:
+        type: string
+      python_version:
+        type: string
+        default: "3.8"
+      git_branch:
+        type: string
+        default: "nightly"
+#        default: "master"
```
Comment (kptkin) [inline .circleci/config.yml:470]: commented out

File: .circleci/config.yml [hunk 6]
```diff
@@ -92,385 +108,894 @@ commands:
...
+      - run:
+          name: "Update gcloud components"
+          command: gcloud --quiet components update
+      - setup_gcloud
+      - get_gke_cluster_credentials:
+          cluster: << parameters.cluster >>
+      - run:
+          name: "Propagate the WANDB_API_KEY environment variable to pod.yaml"
+          command: |
+            sed -i -e 's/WANDB_API_KEY_PLACEHOLDER/'"$WANDB_API_KEY"'/g' << parameters.path >>/<< parameters.pod_config_name >>
```
Comment (kptkin) [inline .circleci/config.yml:610]: interesting solution :)

File: .circleci/config.yml [hunk 7]
```diff
@@ -92,385 +108,894 @@ commands:
...
+            - scheduled_pipeline
+            - << pipeline.trigger_source >>
+        - equal:
+            - "nightly"
+            - << pipeline.schedule.name >>
+    jobs:
+      - slack_notify:
+          name: "slack-notify-on-start"
+          context: slack-secrets
+          message: ":runner: *Nightly run started!*"
```
Comment (kptkin) [inline .circleci/config.yml:691]: should we also add  a link to the run that started?

File: tox.ini [hunk 8]
```diff
@@ -458,3 +472,33 @@ commands =
...
+    export
+    mkdir
+    sh
+commands =
+    mkdir -p test-results
+    wandb login --relogin {env:WANDB_API_KEY}
+    standalone-cpu-py{36,37,38,39,310}: yea -p wandb:mockserver-relay=true -p wandb:mockserver-relay-remote-base-url=https://api.wandb.ai --shard standalone-cpu run {posargs:--all}
+    standalone-gpu-py{36,37,38,39,310}: yea -p wandb:mockserver-relay=true -p wandb:mockserver-relay-remote-base-url=https://api.wandb.ai --shard standalone-gpu run {posargs:--all}
+    standalone-tpu-py{36,37,38,39,310}: yea -p wandb:mockserver-relay=true -p wandb:mockserver-relay-remote-base-url=https://api.wandb.ai --shard standalone-tpu run {posargs:--all}
+    standalone-local-py{36,37,38,39,310}: yea -p wandb:mockserver-bind=0.0.0.0 -p wandb:mockserver-host=__auto__ -p wandb:mockserver-relay=true -p wandb:mockserver-relay-remote-base-url=http://localhost:5000 --shard standalone-cpu run {posargs:--all}
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tox.ini:504]: not a typo as I just wanted to run all the "CPU" tests against local. but if you think we should test more (or less 😅) vs local, i don't have anything against it.

File: standalone_tests/shards/gke_gpu/Dockerfile (hunk 1 above)
Comment (dmitryduev) [reply standalone_tests/shards/gke_gpu/Dockerfile:47]: to get the `standalone_tests` dir...or are you saying we supply that if you install from repo?

File: standalone_tests/basic.yea (hunk 2 above)
Comment (dmitryduev) [reply standalone_tests/basic.yea:5]: yup, sweet, heh?

File: .circleci/config.yml (hunk 3 above)
Comment (dmitryduev) [reply .circleci/config.yml:207]: yea, i can take this out, it's just "notes on the margins" for the upcoming tpu pod.

File: .circleci/config.yml (hunk 4 above)
Comment (dmitryduev) [reply .circleci/config.yml:264]: yea, good idea! will add

File: .circleci/config.yml (hunk 5 above)
Comment (dmitryduev) [reply .circleci/config.yml:470]: yea, I'll swap "nightly" for "master" right before this goes in

File: .circleci/config.yml (hunk 7 above)
Comment (dmitryduev) [reply .circleci/config.yml:691]: too much noise perhaps? there will be now be success messages, but I'll ad a todo anyways

File: tox.ini (hunk 8 above)
[the remark this answers is not in the document]
Comment (kptkin) [reply tox.ini:504]: ah yeah it seems reasonable to me... just wanted to make sure

File: standalone_tests/shards/gke_gpu/Dockerfile (hunk 1 above)
Comment (kptkin) [reply standalone_tests/shards/gke_gpu/Dockerfile:47]: we can have it installed as part of `wandb[test]` or something like this... so we only get it when testing?

File: standalone_tests/basic.yea (hunk 2 above)
Comment (kptkin) [reply standalone_tests/basic.yea:5]: yeah! That's really cool!

File: .circleci/config.yml (hunk 3 above)
Comment (kptkin) [reply .circleci/config.yml:207]: ah that's fine... just keep it :)

File: .circleci/config.yml (hunk 7 above)
Comment (kptkin) [reply .circleci/config.yml:691]: ah no that's fine... one thing I was thinking... don't know if it is supported or how hard to do it... 
that the success/failure message be a thread after the initial run started message ... what do you think?

**Comment by kptkin** [comment]
This is so awesome! Love what you did here!!!!

File: .circleci/config.yml (hunk 7 above)
Comment (dmitryduev) [reply .circleci/config.yml:695]: yea, that'd be awesome, but I don't see this immediately available with the Slack orb

File: .circleci/config.yml [hunk 9]
```diff
@@ -92,30 +108,731 @@ commands:
...
+        type: string
+        default: "pd-ssd"
+      gpu_type:
+        description: "GKE cluster GPU type"
+        type: string
+        default: "nvidia-tesla-t4"
+      gpu_count:
+        description: "GKE cluster GPU count"
+        type: integer
+        default: 1
```
Comment (raubitsj) [inline .circleci/config.yml:177]: Lets bump this to 2

**Review by raubitsj — APPROVED** [review]
nice 🥳

**Pull request merged**